File: example/python/yoy_inflation_capfloor.py
"""Price a year-on-year inflation cap and floor under three distributions.

A `YoYInflationCapFloor` is a strip of optionlets on the year-on-year rate a
`YoYInflationIndex` publishes, one per coupon of an annual inflation leg. It is
built through `MakeYoYInflationCapFloor`, the standard market builder: the core
builder is a consumed-self fluent chain that does not cross the FFI boundary, so
the Python facade takes the whole configuration up front and assembles the chain
inside `build()`.

Three things this shows that no other example does.

* The distribution is chosen by the *constructor*, mirroring C++'s three engine
  classes rather than being passed as an argument: `YoYInflationCapFloorEngine`
  has `black` (lognormal), `unit_displaced` (lognormal in 1 + rate) and
  `bachelier` (normal). On this fixture the last two sit two orders of magnitude
  above Black, so a constructor wired to the wrong distribution cannot hide.
* An engine carries the arguments and results of whatever it last priced, so a
  cap and a floor priced together want one engine each.
* `strike` and `atm_strike` are mutually exclusive and exactly one is required.
  The refusal lands in `build()`, not at the setters, because a fluent setter
  returning `Self` has nowhere to raise from.

The numbers reproduce QuantLib's `inflationcapfloorengines.cpp:452-522`
(`testCachedValue`), as transcribed in
`crates/itofin-py/tests/test_yoy_inflation_capfloor.py`. QuantLib's own
tolerance is 0.02 on a notional of a million, which these only reach if the
curve, the base date, the zero observation lag and the distribution are all
right at once.

Fixture warning. `yoy_inflation_swap.py` next door runs a *different* UK RPI
fixture, and the two must not be crossed. This one files thirty-one published
figures followed by TWO -999.0 sentinels on a BACKWARD-generated schedule, which
carries the index's last fixing date - and with it the curve's base date - out
to 1 August 2007. The leg, the volatility surface and the builder observe ZERO
days of lag while only the fifteen bootstrap helpers observe two months (in C++
this comes of a `CommonVars` member being shadowed by a local in the constructor
body); pricing the leg at two months instead misses every cached value by about
47. The nominal curve is Act/Act (ISDA), not Act/360.

Run it with:

    python example/python/yoy_inflation_capfloor.py
"""

# plugins
# itofin library
from itofin import ItofinError, Settings
from itofin.indexes import CpiInterpolationType, YoYInflationIndex, ZeroInflationIndex
from itofin.instruments import CapFloorType, MakeYoYInflationCapFloor
from itofin.pricingengines import YoYInflationCapFloorEngine
from itofin.quotes import SimpleQuote
from itofin.termstructures import (
    ConstantYoYOptionletVolatility,
    FlatForward,
    PiecewiseYoYInflationCurve,
    YearOnYearInflationSwapHelper,
)
from itofin.time import BusinessDayConvention, Calendar, Date, DateGeneration, DayCounter, Frequency, Period, Schedule

UK = Calendar.united_kingdom()
TODAY = UK.adjust(Date(13, 8, 2007), BusinessDayConvention.Following)
NO_LAG = Period(0, "Days")
HELPER_LAG = Period(2, "Months")
NOMINAL = 1_000_000.0
NOMINAL_RATE = 0.05
STRIKE = 0.0295
VOLATILITY = 0.01
DAY_COUNTER = DayCounter.thirty360_bond_basis()

# Thirty-one published UK RPI figures from January 2005, then two -999.0
# sentinels. Neither sentinel is ever read as a rate - every fixing date here is
# 2008 or later, so the ratio index forecasts off the curve - but they carry the
# curve's base date from June to 1 August 2007, the origin every number below is
# measured from. Dropping them silently moves every optionlet.
FIX_DATA = [
    189.9, 189.9, 189.6, 190.5, 191.6, 192.0, 192.2, 192.2, 192.6, 193.1,
    193.3, 193.6, 194.1, 193.4, 194.2, 195.0, 196.5, 197.7, 198.5, 198.5,
    199.2, 200.1, 200.4, 201.1, 202.7, 201.6, 203.1, 204.4, 205.4, 206.2,
    207.3, -999.0, -999.0,
]

# (maturity, quoted year-on-year swap rate in percent).
YY_DATA = [
    (Date(13, 8, 2008), 2.95),
    (Date(13, 8, 2009), 2.95),
    (Date(13, 8, 2010), 2.93),
    (Date(15, 8, 2011), 2.955),
    (Date(13, 8, 2012), 2.945),
    (Date(13, 8, 2013), 2.985),
    (Date(13, 8, 2014), 3.01),
    (Date(13, 8, 2015), 3.035),
    (Date(13, 8, 2016), 3.055),
    (Date(13, 8, 2017), 3.075),
    (Date(13, 8, 2019), 3.105),
    (Date(15, 8, 2022), 3.135),
    (Date(13, 8, 2027), 3.155),
    (Date(13, 8, 2032), 3.145),
    (Date(13, 8, 2037), 3.145),
]

# inflationcapfloorengines.cpp:452-522, by distribution: (cap, floor, tolerance).
CACHED = {
    "black": (219.452, 314.641, 0.02),
    "unit_displaced": (9114.61, 9209.8, 0.22),
    "bachelier": (8852.4, 8947.59, 0.22),
}


def rpi_schedule() -> Schedule:
    """The dates the RPI figures are filed on: monthly from 1 January 2005 to
    13 August 2007, generated BACKWARD so the schedule walks back from the 13th
    and carries a short front stub. The Python `Schedule` facade defaults to
    Forward, so the rule is passed explicitly; generating forward instead would
    shift every figure by a month."""
    return Schedule(
        Date(1, 1, 2005),
        Date(13, 8, 2007),
        Frequency.Monthly,
        UK,
        BusinessDayConvention.ModifiedFollowing,
        DateGeneration.Backward,
    )


def build_market():
    """One Settings, a ratio year-on-year index over a UK RPI carrying the
    sentinel history, and the 5% nominal curve. The index is returned linked to
    the bootstrapped year-on-year curve."""
    settings = Settings()
    settings.set_evaluation_date(TODAY)

    rpi = ZeroInflationIndex.uk_rpi(settings)
    for date, figure in zip(rpi_schedule().dates(), FIX_DATA):
        rpi.add_fixing(date, figure)

    # A ratio index derives its year-on-year rate from two RPI fixings a year
    # apart; it owns no history of its own, so the fixings go on the underlying.
    index = YoYInflationIndex.from_underlying(rpi)
    nominal = FlatForward(TODAY, NOMINAL_RATE, DayCounter.actual_actual_isda())

    helpers = [
        YearOnYearInflationSwapHelper(
            SimpleQuote(rate / 100.0),
            HELPER_LAG,  # only the helpers observe two months
            maturity,
            UK,
            BusinessDayConvention.ModifiedFollowing,
            DAY_COUNTER,
            index,
            CpiInterpolationType.Flat,
            nominal,
            settings,
        )
        for maturity, rate in YY_DATA
    ]
    curve = PiecewiseYoYInflationCurve(
        TODAY,
        rpi.last_fixing_date(),  # the base date the sentinels carried to August
        YY_DATA[0][1] / 100.0,  # the base year-on-year rate at node zero
        Frequency.Monthly,  # what UK RPI publishes, not the annual leg's
        DAY_COUNTER,
        helpers,
    )
    index.link_to(curve)
    return settings, index, nominal


def surface(settings: Settings) -> ConstantYoYOptionletVolatility:
    """The flat 1% surface the cached values were produced on. Its observation
    lag is zero, like the leg's and unlike the helpers'."""
    return ConstantYoYOptionletVolatility(
        VOLATILITY,
        0,  # settlement days
        UK,
        BusinessDayConvention.ModifiedFollowing,
        DAY_COUNTER,
        NO_LAG,
        Frequency.Annual,
        False,  # index is interpolated
        -1.0,  # min strike
        100.0,  # max strike
        settings,
    )


def cap_floor(settings, index, cap_floor_type, **kwargs):
    """A two-year factory-built cap or floor. The builder's own defaults - an
    annual schedule, a ModifiedFollowing payment roll, 30/360 bond basis, no
    fixing days, every optionlet kept - are exactly the cached fixture's, so
    nothing beyond the strike and the nominal has to be said."""
    return MakeYoYInflationCapFloor(
        cap_floor_type,
        index,
        2,  # length in years
        UK,
        NO_LAG,
        CpiInterpolationType.Flat,
        settings,
        nominal=NOMINAL,
        **kwargs,
    ).build()


def priced(settings, index, nominal, cap_floor_type, distribution) -> float:
    """One instrument, one engine: an engine carries the results of whatever it
    last priced, so a cap and a floor never share one."""
    instrument = cap_floor(settings, index, cap_floor_type, strike=STRIKE)
    engine = getattr(YoYInflationCapFloorEngine, distribution)(index, surface(settings), nominal)
    instrument.set_engine(engine)
    return instrument.npv()


def price_under_each_distribution(settings, index, nominal) -> None:
    print(f"Two-year year-on-year cap and floor struck at {STRIKE:.2%}, notional {NOMINAL:,.0f}, vol {VOLATILITY:.0%}")
    for distribution, (cached_cap, cached_floor, tolerance) in CACHED.items():
        cap = priced(settings, index, nominal, CapFloorType.Cap, distribution)
        floor = priced(settings, index, nominal, CapFloorType.Floor, distribution)
        print(f"\n  {distribution}:")
        print(
            f"    cap   = {cap:10.4f}   cached={cached_cap:>9}   "
            f"|diff|={abs(cap - cached_cap):.4f}  (tol {tolerance})"
        )
        print(
            f"    floor = {floor:10.4f}   cached={cached_floor:>9}   "
            f"|diff|={abs(floor - cached_floor):.4f}  (tol {tolerance})"
        )


def fill_the_strike_at_the_money(settings, index, nominal) -> None:
    """With no strike given, `atm_strike` resolves one off the nominal curve.
    The rate that lands on the leg is the rate the instrument itself would
    reprice at, which `atm_rate` reads back independently.

    Trimming happens before the at-the-money fill, so `as_optionlet` and
    `first_caplet_excluded` would change what an unset strike resolves to: the
    rate repricing whatever survives, not the whole leg's."""
    cap = cap_floor(settings, index, CapFloorType.Cap, atm_strike=nominal)
    struck = cap.cap_rates()[0]

    print("\n\nAt-the-money fill (no strike given, resolved off the nominal curve):")
    print(f"  strike on the leg = {struck:.10f}")
    print(f"  atm_rate()        = {cap.atm_rate(nominal):.10f}   |diff|={abs(struck - cap.atm_rate(nominal)):.2e}")
    print(f"  optionlets        = {cap.coupon_count()}   floor rates on a cap = {cap.floor_rates()}")
    print(f"  runs {cap.start_date()} -> {cap.maturity_date()}")


def strike_and_atm_are_exclusive(settings, index, nominal) -> None:
    """Both together and neither at all are refused, and both refusals surface
    from `build()`."""
    print("\n\nThe strike is required, and exactly one way of giving it:")
    for label, kwargs in [
        ("both a strike and an ATM curve", {"strike": STRIKE, "atm_strike": nominal}),
        ("neither", {}),
    ]:
        try:
            cap_floor(settings, index, CapFloorType.Cap, **kwargs)
        except ItofinError as error:
            print(f"  {label:32} -> {type(error).__name__}: {error}")


def main() -> None:
    settings, index, nominal = build_market()
    print(f"UK RPI last fixing date (the curve's base) = {index.underlying_index().last_fixing_date()}\n")

    price_under_each_distribution(settings, index, nominal)
    fill_the_strike_at_the_money(settings, index, nominal)
    strike_and_atm_are_exclusive(settings, index, nominal)


if __name__ == "__main__":
    main()

File: crates/libitofin/examples/yoy_inflation_capfloor.rs
//! Year-on-year inflation caps and floors priced off a bootstrapped YoY curve.
//!
//! Part A reproduces the six cached values of `test-suite/inflationcapfloor.cpp`
//! `testCachedValue` (`:452-522`): a two-year cap and floor struck at 2.95 % on
//! 1 % volatility, valued under all three optionlet distributions (Black,
//! unit-displaced Black, Bachelier).
//!
//! Part B is `testParity` (`:388-450`): cap - floor reprices the plain
//! `YearOnYearInflationSwap` over the same leg, model-independently.
//!
//! Part C builds a cap through `MakeYoYInflationCapFloor` struck at the money
//! off the nominal curve, and shows the trim-before-fill order: read as a single
//! optionlet, the strike is the surviving coupon's rate rather than the whole
//! leg's.
//!
//! The market is the port's `instrument_oracle` fixture in
//! `crates/libitofin/src/pricingengines/inflation/inflationcapfloorengines.rs`,
//! copied step for step; the builder calls are those of the `tests` module in
//! `crates/libitofin/src/instruments/makeyoyinflationcapfloor.rs`.
//!
//! Two fixture quirks are deliberate, and are what reproduce the cached values:
//! the RPI history overruns by two `-999.0` sentinels, which move the curve's
//! base date from 1 July to 1 August 2007, and the leg observes a ZERO lag while
//! only the bootstrap helpers observe two months.

use libitofin::cashflow::{CashFlow, Leg};
use libitofin::cashflows::{
    CashFlows, YoYInflationCoupon, YoYInflationLeg, YoYOptionletDistribution,
};
use libitofin::errors::QlResult;
use libitofin::handle::{Handle, RelinkableHandle};
use libitofin::indexes::Index; // brings add_fixing / last_fixing_date into scope
use libitofin::indexes::inflation::UkRpi;
use libitofin::indexes::inflationindex::{CpiInterpolationType, InflationIndex, YoYInflationIndex};
use libitofin::instrument::Instrument;
use libitofin::instruments::{
    CapFloorType, MakeYoYInflationCapFloor, SwapType, YearOnYearInflationSwap, YoYInflationCapFloor,
};
use libitofin::interestrate::Compounding;
use libitofin::math::interpolations::linear::Linear;
use libitofin::pricingengine::PricingEngine;
use libitofin::pricingengines::{DiscountingSwapEngine, YoYInflationCapFloorEngine};
use libitofin::quotes::SimpleQuote;
use libitofin::settings::Settings;
use libitofin::shared::{Shared, SharedMut, shared, shared_mut};
use libitofin::termstructures::inflation::inflationhelpers::{
    YearOnYearInflationSwapHelper, YoYInflationHelper,
};
use libitofin::termstructures::inflation::inflationtermstructure::YoYInflationTermStructure;
use libitofin::termstructures::inflation::piecewiseyoyinflationcurve::PiecewiseYoYInflationCurve;
use libitofin::termstructures::volatility::{
    ConstantYoYOptionletVolatility, YoYOptionletVolatilitySurface,
};
use libitofin::termstructures::yields::{FlatForward, Pillar};
use libitofin::termstructures::yieldtermstructure::YieldTermStructure;
use libitofin::time::businessdayconvention::BusinessDayConvention;
use libitofin::time::calendar::Calendar;
use libitofin::time::calendars::unitedkingdom::{Market, UnitedKingdom};
use libitofin::time::date::Month::{August, January};
use libitofin::time::date::{Date, Day, Month, Year};
use libitofin::time::dategenerationrule::DateGeneration;
use libitofin::time::daycounter::DayCounter;
use libitofin::time::daycounters::actualactual::{
    ActualActual, Convention as ActualActualConvention,
};
use libitofin::time::daycounters::thirty360::{Convention, Thirty360};
use libitofin::time::frequency::Frequency;
use libitofin::time::period::Period;
use libitofin::time::schedule::MakeSchedule;
use libitofin::time::timeunit::TimeUnit;
use libitofin::types::{Rate, Real, Volatility};

/// UK RPI, thirty-one real figures then the two `-999.0` sentinels
/// (`inflationcapfloor.cpp:132-137`). See the module docs.
const FIX_DATA: [Real; 33] = [
    189.9, 189.9, 189.6, 190.5, 191.6, 192.0, 192.2, 192.2, 192.6, 193.1, 193.3, 193.6, 194.1,
    193.4, 194.2, 195.0, 196.5, 197.7, 198.5, 198.5, 199.2, 200.1, 200.4, 201.1, 202.7, 201.6,
    203.1, 204.4, 205.4, 206.2, 207.3, -999.0, -999.0,
];

/// The fifteen quoted year-on-year swap rates, in per cent
/// (`inflationcapfloor.cpp:152-168`).
const YY_DATA: [(Day, Month, Year, Real); 15] = [
    (13, August, 2008, 2.95),
    (13, August, 2009, 2.95),
    (13, August, 2010, 2.93),
    (15, August, 2011, 2.955),
    (13, August, 2012, 2.945),
    (13, August, 2013, 2.985),
    (13, August, 2014, 3.01),
    (13, August, 2015, 3.035),
    (13, August, 2016, 3.055),
    (13, August, 2017, 3.075),
    (13, August, 2019, 3.105),
    (15, August, 2022, 3.135),
    (13, August, 2027, 3.155),
    (13, August, 2032, 3.145),
    (13, August, 2037, 3.145),
];

const NOTIONAL: Real = 1_000_000.0;

fn uk() -> Calendar {
    UnitedKingdom::new(Market::Settlement)
}

fn day_counter() -> DayCounter {
    Thirty360::with_convention(Convention::BondBasis)
}

/// The lag the leg, the volatility surface and the parity swap observe: zero.
/// Only the bootstrap helpers observe two months.
fn observation_lag() -> Period {
    Period::new(0, TimeUnit::Days)
}

struct Fixture {
    settings: Shared<Settings<Date>>,
    index: Shared<YoYInflationIndex>,
    nominal: Handle<dyn YieldTermStructure>,
    evaluation_date: Date,
    base_date: Date,
    _curve: Shared<PiecewiseYoYInflationCurve<Linear>>,
    _handle: RelinkableHandle<dyn YoYInflationTermStructure>,
}

/// `CommonVars` (`inflationcapfloor.cpp:109-187`): UK RPI with monthly history,
/// a flat 5 % nominal curve, and a year-on-year curve bootstrapped from the
/// fifteen quotes above.
fn a_bootstrapped_market() -> QlResult<Fixture> {
    let settings = shared(Settings::<Date>::new());
    let evaluation_date = uk().adjust(
        Date::new(13, August, 2007),
        BusinessDayConvention::Following,
    );
    settings.set_evaluation_date(evaluation_date);

    // D11: fixings live on the index, not on a global IndexManager.
    let rpi_schedule = MakeSchedule::new()
        .from(Date::new(1, January, 2005))
        .to(Date::new(13, August, 2007))
        .with_tenor(Period::new(1, TimeUnit::Months))
        .with_calendar(uk())
        .with_convention(BusinessDayConvention::ModifiedFollowing)
        .build();
    let rpi = shared(UkRpi::new(Shared::clone(&settings)));
    for (date, &figure) in rpi_schedule.dates().iter().zip(FIX_DATA.iter()) {
        rpi.add_fixing(*date, figure)?;
    }

    // The YoY index forecasts through an empty relinkable handle, relinked to
    // the curve once it is bootstrapped (D2).
    let handle = RelinkableHandle::<dyn YoYInflationTermStructure>::empty();
    let index = shared(
        YoYInflationIndex::from_underlying(Shared::clone(&rpi))
            .with_term_structure(handle.handle()),
    );
    let nominal = Handle::new(shared(FlatForward::with_rate(
        evaluation_date,
        0.05,
        ActualActual::with_convention(ActualActualConvention::ISDA),
        Compounding::Continuous,
        Frequency::Annual,
    )) as Shared<dyn YieldTermStructure>);

    // Only the helpers see two months (`inflationcapfloor.cpp:150`, `:174`).
    let helper_lag = Period::new(2, TimeUnit::Months);
    let helpers: Vec<Shared<dyn YoYInflationHelper>> = YY_DATA
        .iter()
        .map(|&(day, month, year, rate)| {
            YearOnYearInflationSwapHelper::new(
                Handle::new(shared(SimpleQuote::new(Some(rate / 100.0)))),
                helper_lag,
                Date::new(day, month, year),
                uk(),
                BusinessDayConvention::ModifiedFollowing,
                day_counter(),
                &index,
                CpiInterpolationType::Flat,
                nominal.clone(),
                Pillar::LastRelevantDate,
                Shared::clone(&settings),
            )
            .expect("a well-formed helper") as Shared<dyn YoYInflationHelper>
        })
        .collect();

    // Bootstrap is lazy: the first read off the curve solves it.
    let base_date = rpi.last_fixing_date()?;
    let curve = PiecewiseYoYInflationCurve::<Linear>::new(
        evaluation_date,
        base_date,
        YY_DATA[0].3 / 100.0, // base YoY rate
        index.frequency(),
        day_counter(),
        helpers,
        None, // no seasonality correction
    )?;
    handle.link_to(Shared::clone(&curve) as Shared<dyn YoYInflationTermStructure>);

    Ok(Fixture {
        settings,
        index,
        nominal,
        evaluation_date,
        base_date,
        _curve: curve,
        _handle: handle,
    })
}

/// `makeYoYLeg` (`inflationcapfloor.cpp:190-201`), a plain annual leg.
fn a_yoy_leg(fixture: &Fixture, length: i32) -> QlResult<Vec<Shared<YoYInflationCoupon>>> {
    let start = fixture.evaluation_date;
    let end = uk().advance_by_period(
        start,
        Period::new(length, TimeUnit::Years),
        BusinessDayConvention::Unadjusted,
        false,
    );
    let schedule = MakeSchedule::new()
        .from(start)
        .to(end)
        .with_frequency(Frequency::Annual)
        .with_calendar(uk())
        .with_convention(BusinessDayConvention::Unadjusted)
        .with_termination_date_convention(BusinessDayConvention::Unadjusted)
        .with_rule(DateGeneration::Forward)
        .build();
    YoYInflationLeg::new(
        schedule,
        uk(),
        Shared::clone(&fixture.index),
        observation_lag(),
        CpiInterpolationType::Flat,
    )
    .with_notional(NOTIONAL)
    .with_payment_day_counter(day_counter())
    .with_payment_adjustment(BusinessDayConvention::ModifiedFollowing)
    .coupons()
}

/// `makeEngine` (`inflationcapfloor.cpp:204-241`): a flat optionlet surface
/// under one of the three distributions.
fn an_engine(
    fixture: &Fixture,
    volatility: Volatility,
    distribution: YoYOptionletDistribution,
) -> SharedMut<dyn PricingEngine> {
    let surface = Handle::new(shared(ConstantYoYOptionletVolatility::new(
        volatility,
        0, // settlement days
        uk(),
        BusinessDayConvention::ModifiedFollowing,
        day_counter(),
        observation_lag(),
        Frequency::Annual,
        false, // index is not interpolated
        -1.0,  // minimum strike
        100.0, // maximum strike
        Shared::clone(&fixture.settings),
    )) as Shared<dyn YoYOptionletVolatilitySurface>);
    let index = Shared::clone(&fixture.index);
    let nominal = fixture.nominal.clone();
    let engine = match distribution {
        YoYOptionletDistribution::Black => {
            YoYInflationCapFloorEngine::black(index, surface, nominal)
        }
        YoYOptionletDistribution::UnitDisplaced => {
            YoYInflationCapFloorEngine::unit_displaced(index, surface, nominal)
        }
        YoYOptionletDistribution::Bachelier => {
            YoYInflationCapFloorEngine::bachelier(index, surface, nominal)
        }
    };
    shared_mut(engine) as SharedMut<dyn PricingEngine>
}

/// `makeYoYCapFloor` (`inflationcapfloor.cpp:244-264`): the instrument with its
/// engine already attached.
fn a_cap_floor(
    fixture: &Fixture,
    cap_floor_type: CapFloorType,
    coupons: Vec<Shared<YoYInflationCoupon>>,
    strike: Rate,
    volatility: Volatility,
    distribution: YoYOptionletDistribution,
) -> QlResult<YoYInflationCapFloor> {
    let mut instrument = match cap_floor_type {
        CapFloorType::Floor => {
            YoYInflationCapFloor::floor(coupons, vec![strike], Shared::clone(&fixture.settings))
        }
        _ => YoYInflationCapFloor::cap(coupons, vec![strike], Shared::clone(&fixture.settings)),
    }?;
    instrument
        .base_mut()
        .set_pricing_engine(an_engine(fixture, volatility, distribution));
    Ok(instrument)
}

fn main() -> QlResult<()> {
    let fixture = a_bootstrapped_market()?;
    println!("evaluation date        : {}", fixture.evaluation_date);
    println!("curve base date        : {}", fixture.base_date);
    println!();

    part_a_cached_values(&fixture)?;
    println!();
    part_b_parity(&fixture)?;
    println!();
    part_c_at_the_money_builder(&fixture)?;
    Ok(())
}

// ---------------------------------------------------------------------------
// Part A: the six values QuantLib caches for a 2Y cap and floor.
// ---------------------------------------------------------------------------
fn part_a_cached_values(fixture: &Fixture) -> QlResult<()> {
    println!("== Part A: 2Y cap/floor struck at 2.95 % on 1 % vol vs the cached values ==");
    let strike = 0.0295;
    let coupons = a_yoy_leg(fixture, 2)?;

    // (distribution, cached cap, cached floor) from inflationcapfloor.cpp:474,
    // :493 and :512. QuantLib's own tolerances are 0.02 / 0.22 / 0.22.
    for (distribution, cached_cap, cached_floor) in [
        (YoYOptionletDistribution::Black, 219.452, 314.641),
        (YoYOptionletDistribution::UnitDisplaced, 9114.61, 9209.8),
        (YoYOptionletDistribution::Bachelier, 8852.4, 8947.59),
    ] {
        let mut cap = a_cap_floor(
            fixture,
            CapFloorType::Cap,
            coupons.clone(),
            strike,
            0.01,
            distribution,
        )?;
        let mut floor = a_cap_floor(
            fixture,
            CapFloorType::Floor,
            coupons.clone(),
            strike,
            0.01,
            distribution,
        )?;

        println!(
            "  {distribution:<13?} cap   = {:>10.4}  (cached {cached_cap})",
            cap.npv()?
        );
        println!(
            "  {distribution:<13?} floor = {:>10.4}  (cached {cached_floor})",
            floor.npv()?
        );
    }
    Ok(())
}

// ---------------------------------------------------------------------------
// Part B: cap - floor is the year-on-year swap, whatever the distribution.
// ---------------------------------------------------------------------------
fn part_b_parity(fixture: &Fixture) -> QlResult<()> {
    println!("== Part B: cap - floor == YoY swap (5Y, strike 3 %, vol 15 %) ==");
    let strike = 0.03;
    let volatility = 0.15;
    let length = 5;
    let coupons = a_yoy_leg(fixture, length)?;
    let from = fixture
        .nominal
        .current_link()?
        .reference_date()
        .expect("a fixed-reference curve has one");

    // The swap's own annual schedule over the same span, generated backwards.
    let schedule = MakeSchedule::new()
        .from(from)
        .to(from + Period::new(length, TimeUnit::Years))
        .with_tenor(Period::new(1, TimeUnit::Years))
        .with_calendar(uk())
        .with_convention(BusinessDayConvention::Unadjusted)
        .backwards()
        .build();
    let mut swap = YearOnYearInflationSwap::new(
        SwapType::Payer,
        NOTIONAL,
        schedule.clone(), // fixed schedule
        strike,           // fixed rate
        day_counter(),    // fixed day counter
        schedule,         // yoy schedule
        Shared::clone(&fixture.index),
        observation_lag(),
        CpiInterpolationType::Flat,
        0.0,           // spread
        day_counter(), // yoy day counter
        uk(),
        BusinessDayConvention::ModifiedFollowing,
        Shared::clone(&fixture.settings),
    )?;
    swap.base_mut()
        .set_pricing_engine(shared_mut(DiscountingSwapEngine::new(
            fixture.nominal.clone(),
            None,
            None,
            None,
            Shared::clone(&fixture.settings),
        )) as SharedMut<dyn PricingEngine>);
    let swap_npv = swap.npv()?;

    for distribution in [
        YoYOptionletDistribution::Black,
        YoYOptionletDistribution::UnitDisplaced,
        YoYOptionletDistribution::Bachelier,
    ] {
        let mut cap = a_cap_floor(
            fixture,
            CapFloorType::Cap,
            coupons.clone(),
            strike,
            volatility,
            distribution,
        )?;
        let mut floor = a_cap_floor(
            fixture,
            CapFloorType::Floor,
            coupons.clone(),
            strike,
            volatility,
            distribution,
        )?;
        let parity = (cap.npv()? - floor.npv()?) - swap_npv;
        println!("  {distribution:<13?} cap - floor - swap = {parity:.3e}");
    }
    println!("  swap NPV = {swap_npv:.6}");
    Ok(())
}

// ---------------------------------------------------------------------------
// Part C: MakeYoYInflationCapFloor, at-the-money strikes, and atm_rate.
// ---------------------------------------------------------------------------
fn part_c_at_the_money_builder(fixture: &Fixture) -> QlResult<()> {
    println!("== Part C: MakeYoYInflationCapFloor at the money ==");
    let curve = fixture.nominal.current_link()?;

    let builder = || {
        MakeYoYInflationCapFloor::new(
            CapFloorType::Cap,
            Shared::clone(&fixture.index),
            5, // length in years
            uk(),
            observation_lag(),
            CpiInterpolationType::Flat,
            Shared::clone(&fixture.settings),
        )
        .with_nominal(NOTIONAL)
    };

    // A strike left unset is filled at the money off the nominal curve: the rate
    // that reprices the leg the builder just laid down.
    let whole_leg = builder().with_atm_strike(fixture.nominal.clone()).build()?;
    println!(
        "  5Y cap, atm strike       = {:.10}",
        whole_leg.cap_rates()[0]
    );
    println!(
        "  5Y cap, atm_rate()       = {:.10}",
        whole_leg.atm_rate(curve.as_ref())?
    );

    // atm_rate is CashFlows::atm_rate over the instrument's own leg, off the
    // curve's reference date.
    let leg: Leg = whole_leg
        .yoy_leg()
        .iter()
        .map(|coupon| Shared::clone(coupon) as Shared<dyn CashFlow>)
        .collect();
    let by_hand = CashFlows::atm_rate(
        &leg,
        curve.as_ref(),
        &fixture.settings,
        Some(false), // include settlement date flows
        Some(curve.reference_date().expect("a reference date")),
        None, // npv date
        None, // target npv
    )?;
    println!("  the same by hand         = {by_hand:.10}");

    // as_optionlet keeps only the last coupon, and the trim happens BEFORE the
    // at-the-money fill: the strike is that coupon's rate, not the whole leg's.
    // This market is near flat, so the two sit under ten basis points apart; on
    // a curve with real slope the gap runs to a couple of hundred.
    let optionlet = builder()
        .as_optionlet(true)
        .with_atm_strike(fixture.nominal.clone())
        .build()?;
    println!(
        "  last-coupon optionlet    = {:.10}  ({} coupon)",
        optionlet.cap_rates()[0],
        optionlet.yoy_leg().len()
    );
    Ok(())
}
